py/server/tests/test_udf_numpy_args.py

```diff
@@ -0,0 +1,384 @@
+            with self.assertRaises(DHError) as cm:
+                t2 = t.update(["X1 = f2(null, Y )"])
+
+            def f21(p1: Union[np.ndarray[np.datetime64], None], p2=None) -> bool:
+                return bool(len(p1)) if p1 is not None else False
+            t2 = t.update(["X1 = f21(null, Y)"])
+            self.assertEqual(3, t2.to_string().count("false"))
+
+        with self.subTest("boolean"):
+            def f3(p1: np.ndarray[np.bool_], p2=None) -> bool:
+                return bool(len(p1))
+
+            t = empty_table(10).update(["X = i % 3", "Y = i % 2 == 0? true : null"]).group_by("X")
+            with self.assertRaises(DHError) as cm:
+                t1 = t.update(["X1 = f3(Y)"])
+
+            t = empty_table(10).update(["X = i % 3", "Y = i % 2 == 0? true : false"]).group_by("X")
+            t1 = t.update(["X1 = f3(Y)"])
+            self.assertEqual(t1.columns[2].data_type, dtypes.bool_)
+            with self.assertRaises(DHError) as cm:
+                t2 = t.update(["X1 = f3(null, Y )"])
+
+            def f31(p1: Optional[np.ndarray[bool]], p2=None) -> bool:
+                return bool(len(p1)) if p1 is not None else False
+            t2 = t.update(["X1 = f31(null, Y)"])
+            self.assertEqual(3, t2.to_string("X1").count("false"))
+
+    def test_str_bool_datetime_scalar(self):
+        with self.subTest("str"):
+            def f1(p1: str, p2=None) -> bool:
+                return p1 is None
+
+            t = empty_table(10).update(["X = i % 3", "Y = i % 2 == 0? `deephaven`: null"])
+            t1 = t.update(["X1 = f1(Y)"])
+            self.assertEqual(t1.columns[2].data_type, dtypes.bool_)
+            self.assertEqual(5, t1.to_string().count("true"))
+
+            def f11(p1: Union[str, None], p2=None) -> bool:
+                return p1 is None
+            t2 = t.update(["X1 = f11(Y)"])
+            self.assertEqual(5, t2.to_string().count("false"))
+
+        with self.subTest("datetime"):
+            def f2(p1: np.datetime64, p2=None) -> bool:
+                return p1 is None
+
+            t = empty_table(10).update(["X = i % 3", "Y = i % 2 == 0? now() : null"])
+            t1 = t.update(["X1 = f2(Y)"])
+            self.assertEqual(t1.columns[2].data_type, dtypes.bool_)
+            self.assertEqual(5, t1.to_string().count("true"))
+
+            def f21(p1: Union[np.datetime64, None], p2=None) -> bool:
+                return p1 is None
+            t2 = t.update(["X1 = f21(Y)"])
+            self.assertEqual(5, t2.to_string().count("false"))
+
+        with self.subTest("boolean"):
+            def f3(p1: np.bool_, p2=None) -> bool:
+                return p1 is None
+
+            t = empty_table(10).update(["X = i % 3", "Y = i % 2 == 0? true : null"])
+            t1 = t.update(["X1 = f3(Y)"])
+            self.assertEqual(t1.columns[2].data_type, dtypes.bool_)
+            self.assertEqual(5, t1.to_string("X1").count("false"))
```

Comment:
This seems to violate the annotation. Are null inputs allowed for object types if they are not optional?